java/ql/lib/semmle/code/java/security/TemplateInjectionQuery.qll

```diff
@@ -0,0 +1,40 @@
+/** Provides a taint tracking configuration for server-side template injection (SST) vulnerabilities */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.security.TemplateInjection
+
+/** A taint tracking configuration to reason about server-side template injection (SST) vulnerabilities */
+class TemplateInjectionFlowConfig extends TaintTracking::Configuration {
+  TemplateInjectionFlowConfig() { this = "TemplateInjectionFlowConfig" }
+
+  override predicate isSource(DataFlow::Node source) { this.isSource(source, _) }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    source.(TemplateInjectionSource).hasState(state)
+  }
+
+  override predicate isSink(DataFlow::Node sink) { this.isSink(sink, _) }
+
+  override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+    sink.(TemplateInjectionSink).hasState(state)
+  }
+
+  override predicate isSanitizer(DataFlow::Node sanitizer) { this.isSanitizer(sanitizer, _) }
+
+  override predicate isSanitizer(DataFlow::Node sanitizer, DataFlow::FlowState state) {
+    sanitizer.(TemplateInjectionSanitizer).hasState(state)
+  }
```

Comment:
The case with sanitizers is different, although there's also something wrong here. The stateless sanitizer predicate applies to _all_ states, so if you want both regular sanitizers and stateful sanitizers to be configurable, then they'll need a separate extension point.

Reply:
Hm, I think I'm missing something. In the current implementation, the stateless `isSanitizer` actually applies to _all_ states as intended, doesn't it? Well, to any `state` where `any(TemplateInjectionSanitizer).hasState(state)` holds, but I think that's functionally the same (?).

Should we just change the default state of the sanitizer to be `any()`:

```
abstract class TemplateInjectionSanitizer extends DataFlow::Node {
  /** Holds if this sanitizer has the specified `state`. */
  predicate hasState(DataFlow::FlowState state) { any() }
}
```

so that its extensions apply to all states by default, and then users can create an extensions that overrides `hasState` to create state-specific sanitizers?

Reply:
`FlowState` is an infinite type, so making changing the code to `any()` like that is equivalent to making it `abstract` (if it even compiles, which I doubt).
The problem is that if you currently add a sanitizer by extending `TemplateInjectionSanitizer` then it's going to apply to all states regardless of whether you override `hasState` or not.
And there's really no way of fixing that with just a single predicate extension point to represent both state-oblivious sanitizers and state-based sanitizers, since the former can't be represented as part of a binary predicate when the set of possible states is infinite.